pkg/store/cache/matchers_cache.go

```diff
@@ -0,0 +1,193 @@
+	c.metrics.requestsTotal.Inc()
+	if item, ok := c.cache.Get(key); ok {
+		c.metrics.hitsTotal.Inc()
+		return item, nil
+	}
+
+	v, err, _ := c.sf.Do(key.String(), func() (interface{}, error) {
+		if item, ok := c.cache.Get(key); ok {
+			c.metrics.hitsTotal.Inc()
+			return item, nil
+		}
+
+		item, err := newItem(key)
+		if err != nil {
+			return nil, err
+		}
+		c.cache.Add(key, item)
+		c.metrics.numItems.Set(float64(c.cache.Len()))
+		return item, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return v.(*labels.Matcher), nil
+}
+
+func (c *LruMatchersCache) onEvict(_ ConversionLabelMatcher, _ *labels.Matcher) {
+	c.metrics.evicted.Inc()
+	c.metrics.numItems.Set(float64(c.cache.Len()))
+}
+
+type matcherCacheMetrics struct {
+	requestsTotal prometheus.Counter
+	hitsTotal     prometheus.Counter
+	numItems      prometheus.Gauge
+	maxItems      prometheus.Gauge
+	evicted       prometheus.Counter
+}
+
+func newMatcherCacheMetrics(reg prometheus.Registerer) *matcherCacheMetrics {
+	return &matcherCacheMetrics{
+		requestsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_matchers_cache_requests_total",
+			Help: "Total number of cache requests for series matchers",
+		}),
+		hitsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_matchers_cache_hits_total",
+			Help: "Total number of cache hits for series matchers",
+		}),
+		numItems: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "thanos_matchers_cache_items",
+			Help: "Total number of cached items",
+		}),
+		maxItems: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "thanos_matchers_cache_max_items",
+			Help: "Maximum number of items that can be cached",
+		}),
+		evicted: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_matchers_cache_evicted_total",
+			Help: "Total number of items evicted from the cache",
+		}),
+	}
+}
+
+// MatchersToPromMatchersCached returns Prometheus matchers from proto matchers.
+// Works analogously to MatchersToPromMatchers but uses cache to avoid unnecessary allocations and conversions.
+// NOTE: It (can) allocate memory.
+func MatchersToPromMatchersCached(cache MatchersCache, ms ...storepb.LabelMatcher) ([]*labels.Matcher, error) {
+	res := make([]*labels.Matcher, 0, len(ms))
+	for i := range ms {
+		pm, err := cache.GetOrSet(&ms[i], MatcherToPromMatcher)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, pm)
+	}
+	return res, nil
+}
+
+func MatcherToPromMatcher(m ConversionLabelMatcher) (*labels.Matcher, error) {
+	mi, ok := m.(*storepb.LabelMatcher)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("invalid matcher type. Got: %T", m))
+	}
+
+	return storepb.MatcherToPromMatcher(*mi)
+}
+
+// ConversionLabelMatcher is a common interface for the Prometheus and Thanos label matchers.
+type ConversionLabelMatcher interface {
+	String() string
+	GetName() string
+	GetType() prompb.LabelMatcher_Type
```

Comment:
Do we need this method? Askin because the return type `prompb.LabelMatcher_Type` is a Thanos package not Prometheus match type